File: src/dotnet/cdk/Constructs/ResiliantQueue.cs
using Amazon.CDK.AWS.SQS;
using Constructs;

namespace ServerlessGettingStarted.CDK.Constructs;

public record ResilientQueueProps(string QueueName, string Env);

public class ResilientQueue : Construct
{
    public IQueue Queue { get; }
    public IQueue Dlq { get; }

    public ResilientQueue(Construct scope, string id, ResilientQueueProps props) : base(scope, id)
    {
        this.Dlq = new Queue(this, $"{props.QueueName}DLQ-{props.Env}", new QueueProps()
        {
            QueueName = $"{props.QueueName}DLQ-{props.Env}"
        });
        this.Queue = new Queue(this, $"{props.QueueName}-{props.Env}", new QueueProps()
        {
            QueueName = $"{props.QueueName}-{props.Env}",
            DeadLetterQueue = new DeadLetterQueue()
            {
                Queue = this.Dlq,
                MaxReceiveCount = 3
            }
        });
    }
}

File: src/go/cdk/sharedConstructs/resiliantQueue.go
package sharedconstructs

import (
	sharedprops "cdk/shared"

	"github.com/aws/aws-cdk-go/awscdk/v2/awssqs"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type ResiliantQueueProps struct {
	QueueName   string
	SharedProps sharedprops.SharedProps
}

type ResiliantQueue struct {
	Queue awssqs.IQueue
	DLQ   awssqs.IQueue
}

func NewResiliantQueue(scope constructs.Construct, id string, props *ResiliantQueueProps) ResiliantQueue {
	resiliantQueue := ResiliantQueue{}

	resiliantQueue.DLQ = awssqs.NewQueue(scope, jsii.Sprintf("%sDLQ-%s", props.QueueName, props.SharedProps.Env), &awssqs.QueueProps{
		QueueName: jsii.Sprintf("%sDLQ-%s", props.QueueName, props.SharedProps.Env),
	})

	resiliantQueue.Queue = awssqs.NewQueue(scope, jsii.Sprintf("%s-%s", props.QueueName, props.SharedProps.Env), &awssqs.QueueProps{
		QueueName: jsii.Sprintf("%s-%s", props.QueueName, props.SharedProps.Env),
		DeadLetterQueue: &awssqs.DeadLetterQueue{
			MaxReceiveCount: jsii.Number(3),
			Queue:           resiliantQueue.DLQ,
		},
	})

	return resiliantQueue
}
